modules/hostdb/entries.go

```diff
@@ -1,42 +1,153 @@
 package hostdb
 
 import (
+	"crypto/rand"
+	"errors"
+	"math/big"
+
 	"github.com/NebulousLabs/Sia/consensus"
 	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/network"
 )
 
-// TODO: Add a bunch of different ways to arrive at weight, which can each be
-// chosen according to the need at hand. This might also require having a bunch
-// of different weights at each node in the tree.
-
-// entryWeight determines the weight of a specific host, which is:
-//
-//		Freeze * Collateral / sqrt(Price).
-//
-// Freeze has to be linear, because any non-linear freeze will invite sybil
-// attacks.
-//
-// For now, collateral is also linear because an increased collateral means
-// increased risk for the host. (Freeze on the other hand has no risk.) It
-// might be better to make collateral grow sublinearly, such as taking
-// sqrt(collateral) or collateral^(4/5).
-//
-// We take the square of the price to heavily emphasize hosts that have a low
-// price. This is also a bit simplistic however, because we're not sure what
-// the host might be charging for bandwidth.
-func entryWeight(entry modules.HostEntry) consensus.Currency {
-	// Catch a divide by 0 error, and let all hosts have at least some weight.
-	//
-	// TODO: Perhaps there's a better way to do this.
-	if entry.Price.Sign() <= 0 {
-		entry.Price = consensus.NewCurrency64(1)
-	}
-	if entry.Collateral.Sign() <= 0 {
-		entry.Collateral = consensus.NewCurrency64(1)
-	}
-	if entry.Freeze.Sign() <= 0 {
-		entry.Freeze = consensus.NewCurrency64(1)
-	}
-
-	return entry.Freeze.Mul(entry.Collateral).Div(entry.Price.Sqrt())
+// entryWeight returns the weight of a host entry according to the internal
+// metrics of the HostDB. Currently, that means using the weight (10^29 *
+// entry.Collateral / (entry.Price)^2), where entry.Collateral is adjusted to
+// be at most twice the price and at least half the price.
+func entryWeight(entry modules.HostEntry) (weight consensus.Currency) {
+	// Find the collateral by enforcing a min and max collateral. The minimum collateral
+	maxCollateral := entry.Price.Mul(consensus.NewCurrency64(2))
+	minCollateral := entry.Price.Div(consensus.NewCurrency64(2))
+	var collateral consensus.Currency
+	if maxCollateral.Cmp(entry.Collateral) < 0 {
+		collateral = maxCollateral
+	} else if minCollateral.Cmp(entry.Collateral) > 0 {
+		collateral = minCollateral
+	} else {
+		collateral = entry.Collateral
+	}
```

Comment:
docstring is incomplete.
I would write this as:
```go
collateral := entry.Collateral
// clamp collateral to between 0.5x and 2x the price
if collateral.Cmp(entry.Price.MulFloat(2)) > 0 {
    collateral = entry.Price.MulFloat(2)
} else if collateral.Cmp(entry.Price.MulFloat(0.5)) < 0 {
    collateral = entry.Price.MulFloat(0.5)
}
```
(Might be cleaner to write `if maxCollateral := entry.Price.MulFloat(2); collateral.Cmp(maxCollateral) > 0 {` to avoid recalculation, but this function doesn't need to be terribly efficient anyway.)